subsys/bluetooth/host/id.c

```diff
@@ -656,7 +656,7 @@ static int hci_id_add(uint8_t id, const bt_addr_le_t *addr, uint8_t peer_irk[16]
 	memcpy(cp->peer_irk, peer_irk, 16);
 
 #if defined(CONFIG_BT_PRIVACY)
-	memcpy(cp->local_irk, bt_dev.irk[id], 16);
+	memcpy(cp->local_irk, &bt_dev.irk[id], 16);
```

Comment:
Is this change really necessary? 

As far as I can tell `bt_dev.irk[id]` should work. 

If you want to carry on with this change, please also add (void) to the return:
```suggestion
	(void)memcpy(cp->local_irk, &bt_dev.irk[id], 16);
```

Reply:
I'm also surprised that we are still using a literal int for the size, instead of the `sizeof` function or even just a macro, but that's outside the scope of this PR I feel.

Reply:
Yes, but look at the commit message. The types of `bt_dev.irk[id]` and `&bt_dev.irk[id]` are different. The latter is `uint8_t*` which is really what we want here.

Reply:
The `irk` is defined as 
```
uint8_t			irk[CONFIG_BT_ID_MAX][16];
```
`bt_dev.irk[id]` is then a `uint8_t arr[16]` type, right? Which is the same type as `cp->local_irk`. 

Based on the commit message, shouldn't it have been `&bt_dev.irk[id][0]` as the commit message says: 

> Provide a pointer to the first element in the IRK in instead of the pointer to the array.

Reply:
> Based on the commit message, shouldn't it have been &bt_dev.irk[id][0] as the commit message says:
>> Provide a pointer to the first element in the IRK in instead of the pointer to the array.

Yes. The commit message is inconsistent with the change. Needs fix.

I as well think the code has semantic problems now.

#### Rationale for semantic problem

I think that `dst` and `src` in `memcpy` should be of the same type before array-to-pointer-decay for clarity.

##### The proposed fix does not follow this rule.
```c
memcpy(cp->local_irk, &bt_dev.irk[id], 16);
```
`dst` has type `uint8_t [16]` while `src` has `uint8_t (*)[16]`.

##### Semantically, there are two sensible options.

###### Object copy.
Code is ignorant of the type and size of what is copied. This is the analog of using `memcpy` to copy an `int`. (Both `src`/`dst` are typed `uint8_t (*)[16]`).
```c
memcpy(&cp->local_irk, &bt_dev.irk[id], sizeof(cp->local_irk));
```

###### Array copy with explicit decay.
We explicitly decay the array to pointer to its first element with `&arr[0]`. (Both `src`/`dst` are typed `uint8_t *`).
```c
memcpy(&cp->local_irk[0], &bt_dev.irk[id][0], 16);
```
This has a slightly different meaning. This code insist on copying an array. If `cp->local_irk` and/or `bt_dev.irk[id]` are refactored to pointers, this code will still perform an array copy. The length of the array can of course instead be computed with e.g. `sizeof`.

Reply:
@alwa-nordic Do you have a suggestion (both change and commit message). keep in mind that bt_dev.irk[id] is accessed in multiple places, so using `&bt_dev.irk[id]` would be consistent with the other uses. Coverity hasn't complained about any of those so far.

Reply:
`&bt_dev.irk[id]` is fine by me. But I prefer the array-semantics in this particular case. It would be consistent with the use of `memset` in this function.

If we go for `&bt_dev.irk[id]` I think we should add an explicit reference `&` to `dst`, like in my previous comment. I don't know how Coverity will react to this. But would be sensible code to me. It's as if using a macro like
```c
#define memcpy_object(dst, src) memcpy(&(dst), &(src), sizeof(dst))

struct X a; struct X b;
memcpy_object(a, b);
```

As to the commit message.. I realize now that it's unclear to me what exactly tripped up Coverity here. I'm unable to open the Coverity report myself. The github bug report says "Out-of-bounds access", but I don't see any. Maybe it's because we don't check `id < CONFIG_BT_ID_MAX`?

Reply:
See https://stackoverflow.com/questions/2528318/how-come-an-arrays-address-is-equal-to-its-value-in-c/2528328#2528328

Reply:
Thanks for all the comments.
I think we should add the `&` for consistency with other `bt_dev.irk` uses with `memcpy` and then mark the Coverity issue as false positive, since adding an assert here would not be very useful (we index by `id` in several other places).

Agreed @alwa-nordic and @Thalley ?

Reply:
> The github bug report says "Out-of-bounds access", but I don't see any. Maybe it's because we don't check id < CONFIG_BT_ID_MAX

I've been looking into the ways that the id is assigned and I believe with the other changes in this PR that they should all have been handled.
The `bt_id_*` function checks the input.
The connection objects provide the ID, peripheral should be handled by validating the advertiser params, and the central can only be BT_ID_DEFAULT.

Reply:
As @carlescufi pointed out, the change has not functional change, and if it shuts up Coverity it'll be OK. 

I agree with @alwa-nordic that semantically the line is now a bit odd, though. 

In any case, I'll rather get this fixed and then worry about semantics later. 

@joerchan I'll approve this PR, but could you please also add a `(void`) before the memcpy as well?